util/src/io/util.rs

```diff
@@ -0,0 +1,216 @@
+use crate::io::{self, ErrorKind, Initializer, Read, Write};
+use core::{fmt, mem::MaybeUninit};
+
+/// Copies the entire contents of a reader into a writer.
+///
+/// This function will continuously read data from `reader` and then
+/// write it into `writer` in a streaming fashion until `reader`
+/// returns EOF.
+///
+/// On success, the total number of bytes that were copied from
+/// `reader` to `writer` is returned.
+///
+/// # Errors
+///
+/// This function will return an error immediately if any call to `read` or
+/// `write` returns an error. All instances of `ErrorKind::Interrupted` are
+/// handled by this function and the underlying operation is retried.
+pub fn copy<R: ?Sized, W: ?Sized>(reader: &mut R, writer: &mut W) -> io::Result<u64>
+where
+    R: Read,
+    W: Write,
+{
+    let mut buf = MaybeUninit::<[u8; super::DEFAULT_BUF_SIZE]>::uninit();
+    // FIXME(eliza): the stdlib has a scary comment that says we haven't decided
+    // if this is okay or not...
```

Comment:
Yeah, this has to do with the initialization thing I mention in an earlier comment. The `read` method takes a `&mut [u8]` for historical reasons, and IIRC it hasn't been decided yet whether or not passing uninitialized memory as a `&mut [u8]` is OK.

The `Initializer` mechanism is effectively being used here as a way of encoding a property of a method not encoded by the type system.